File: android/measure/src/main/java/sh/measure/android/tracing/SpanCollector.kt
package sh.measure.android.tracing

import java.util.concurrent.atomic.AtomicBoolean

internal class SpanCollector(private val tracer: Tracer) {
    private val isEnabled = AtomicBoolean(false)

    fun register() {
        isEnabled.set(true)
    }

    fun unregister() {
        isEnabled.set(false)
    }

    fun getTraceParentHeaderValue(span: Span): String {
        return tracer.getTraceParentHeaderValue(span)
    }

    fun getTraceParentHeaderKey(): String {
        return tracer.getTraceParentHeaderKey()
    }

    fun createSpan(name: String): SpanBuilder? {
        if (!isEnabled.get()) {
            return null
        }
        return tracer.spanBuilder(name)
    }

    fun startSpan(name: String, setNoParent: Boolean = false, timestamp: Long? = null): Span {
        if (!isEnabled.get()) {
            return Span.invalid()
        }

        val spanBuilder = tracer.spanBuilder(name)
        if (setNoParent) {
            spanBuilder.setNoParent()
        }
        if (timestamp != null) {
            spanBuilder.startSpan(timestamp)
        }
        return spanBuilder.startSpan()
    }

    fun getCurrentSpan(): Span? {
        if (!isEnabled.get()) {
            return null
        }
        return Span.current()
    }
}

File: ios/Sources/MeasureSDK/Swift/Tracing/SpanCollector.swift
//
//  SpanCollector.swift
//  Measure
//
//  Created by Adwin Ross on 10/04/25.
//

import Foundation

protocol SpanCollector {
    func enable()
    func disabled()
    func getTraceParentHeaderValue(for span: Span) -> String
    func getTraceParentHeaderKey() -> String
    func createSpan(name: String) -> SpanBuilder?
    func startSpan(name: String, timestamp: Int64?) -> Span
}

final class BaseSpanCollector: SpanCollector {
    private let tracer: Tracer
    private var isEnabled = AtomicBool(false)

    init(tracer: Tracer) {
        self.tracer = tracer
    }

    func enable() {
        isEnabled.set(true)
    }

    func disabled() {
        isEnabled.set(false)
    }

    func getTraceParentHeaderValue(for span: Span) -> String {
        return tracer.getTraceParentHeaderValue(for: span)
    }

    func getTraceParentHeaderKey() -> String {
        return tracer.getTraceParentHeaderKey()
    }

    func createSpan(name: String) -> SpanBuilder? {
        guard isEnabled.get() else {
            return nil
        }
        return tracer.spanBuilder(name: name)
    }

    func startSpan(name: String, timestamp: Int64?) -> Span {
        guard isEnabled.get() else {
            return InvalidSpan()
        }

        let spanBuilder = tracer.spanBuilder(name: name)
        if let timestampMs = timestamp {
            return spanBuilder.startSpan(timestampMs)
        }
        return spanBuilder.startSpan()
    }
}
